algoliasearch-common/src/main/java/com/algolia/search/objects/FacetFilters.java

```diff
@@ -0,0 +1,76 @@
+package com.algolia.search.objects;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.*;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.List;
+
+@JsonDeserialize(using = FacetFiltersJsonDeserializer.class)
+@JsonSerialize(using = FacetFiltersJsonSerializer.class)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public abstract class FacetFilters implements Serializable {
+
+  public static FacetFilters ofList(List<String> filters) {
+    return new FacetFiltersAsListOfString(filters);
+  }
+
+  public static FacetFilters ofListOfList(List<List<String>> filters) {
+    return new FacetFiltersAsListOfList(filters);
+  }
+
+  @JsonIgnore
+  public abstract Object getInsideValue();
+}
+
+class FacetFiltersJsonDeserializer extends JsonDeserializer {
+
+  @Override
+  public Object deserialize(JsonParser p, DeserializationContext ctxt)
+      throws IOException, JsonProcessingException {
+    List list = p.readValueAs(List.class);
+    if (list.isEmpty()) {
+      return new FacetFiltersAsListOfString();
+    }
+    if (list.get(0) instanceof String) {
+      return new FacetFiltersAsListOfString(list);
+    }
+    return new FacetFiltersAsListOfList(list);
```

Comment:
As the `FacetFiltersAsListOfList` is expecting a `List<List<String>>`, I supposed this would throw if `list` was not of type `List<List<String>>`, hence validating the expected behaviour. Does it look wrong to you?

Otherwise, do you want me to check that the first element is of type `List<String>` and otherwise throw a custom exception?

Reply:
I would throw a meaning full exception instead on relying on what the JVM will generate